internal/pkg/agent/application/upgrade/step_mark.go

```diff
@@ -277,11 +278,50 @@ func CleanMarker(log *logger.Logger, dataDirPath string) error {
 // LoadMarker loads the update marker. If the file does not exist it returns nil
 // and no error.
 func LoadMarker(dataDirPath string) (*UpdateMarker, error) {
-	return loadMarker(markerFilePath(dataDirPath))
+	return loadMarker(markerFilePath(dataDirPath), true)
 }
 
-func loadMarker(markerFile string) (*UpdateMarker, error) {
-	markerBytes, err := readMarkerFile(markerFile)
+func UpdateMarkerFile(markerFile string, opts ...func(*UpdateMarker)) error {
+	fileLock, err := newMarkerFileLocker(markerFile)
+	if err != nil {
+		return fmt.Errorf("creating update marker locker for reading: %w", err)
+	}
+	err = fileLock.Lock()
+	if err != nil {
+		return fmt.Errorf("locking update marker file %q for reading: %w", markerFile, err)
+	}
+	defer func(fileLock Locker) {
+		errUnlock := fileLock.Unlock()
+		if errUnlock != nil {
+			err = goerrors.Join(err, fmt.Errorf("unlocking marker file after reading: %w", errUnlock))
+		}
+	}(fileLock)
+
+	marker, err := loadMarker(markerFile, false)
```

Comment:
Are we shadowing err here ? (thinking of the `errors.Join()` in line 296 mostly)